modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala

```diff
@@ -80,153 +84,194 @@ private[smithy4s] class SchemaVisitorJCodec(
         def encodeKey(x: String, out: JsonWriter): Unit = out.writeKey(x)
       }
 
-    val int: JCodec[Int] =
-      new JCodec[Int] {
-        def expecting: String = "int"
+    private abstract class NumericJCodec[A] extends JCodec[A] {
+      def decodeJsonNumber(cursor: Cursor, in: JsonReader): A
+      // Allows numerics to be received as JSON Strings
+      def decodeJsonString(cursor: Cursor, in: JsonReader): A
 
-        def decodeValue(cursor: Cursor, in: JsonReader): Int = in.readInt()
+      final def decodeValue(cursor: Cursor, in: JsonReader): A =
+        if (allowJsonStringNumerics) {
```

Comment:
It's just one if statement, but I think this could be moved off the hot path?

Reply:
Good question. In this case, it's hard : the problem is that due to the imperative nature of jsoniter, the logic is not compositional. If we want to get checks on the hotpath at all cost, that in an ideal world we'd have one class per combination of flags 

1. disallowed strings but allowed infinity 
2. allowed strings and allowed infinity 
3. disallowed strings and disallowed infinity 
4. allowed strings but disallowed infinity 

Right now I'm accepting this boolean check as a compromise to avoid having to repeat inherently error-prone bits of logic in 4 different implementations. I'm not too happy about it but I'd rather live with it.

Reply:
I think I'm following, for some reason I thought that you could also do at the class level something like:

```
private val decodeV: (Cursor, JsonReader) => A = if (allowJsonStringNumerics) {
  (cursor, in) => // ...
} else {
  (cursor, in) => // ...
}

final def decodeValue(cursor: Cursor, in: JsonReader): A = this.decodeV(cursor, in)
```

Reply:
also there's the question of : is a method dispatch faster than a boolean check in this context ? not sure

Reply:
Yeah good point, it probably isn't worth worrying about, was mostly just asking to clarify my understanding